cpp/src/merge/merge.cu

```diff
@@ -0,0 +1,381 @@
+namespace experimental { 
+namespace detail {
+
+std::unique_ptr<cudf::table> merge(table_view const& left_table,
+                                   table_view const& right_table,
+                                   std::vector<cudf::size_type> const& key_cols,
+                                   std::vector<order_by_type> const& asc_desc,
+                                   bool nulls_are_smallest,
+                                   cudaStream_t stream = 0) {
+    CUDF_EXPECTS(left_table.num_columns() == right_table.num_columns(), "Mismatched number of columns");
+    if (left_table.num_columns() == 0) {
+        return cudf::empty_like(left_table);
+    }
+
+    // TODO: replace / drop;
+    //no replacement, yet:
+    //{
+    //proposal for a replacement: bool have_same_types(table_view const& lhs, table_view const& rhs);
+    //(btw, issues like that are the reason for 1st refactoring and then creating a 2nd PR)
+    //
+    std::vector<gdf_dtype> left_table_dtypes = cudf::column_dtypes(left_table);
+    std::vector<gdf_dtype> right_table_dtypes = cudf::column_dtypes(right_table);
+    CUDF_EXPECTS(std::equal(left_table_dtypes.cbegin(), left_table_dtypes.cend(), right_table_dtypes.cbegin(), right_table_dtypes.cend()), "Mismatched column dtypes");
+    //}
+    
+    CUDF_EXPECTS(key_cols.size() > 0, "Empty key_cols");
+    CUDF_EXPECTS(key_cols.size() <= static_cast<size_t>(left_table.num_columns()), "Too many values in key_cols");
+    CUDF_EXPECTS(asc_desc.size() > 0, "Empty asc_desc");
+    CUDF_EXPECTS(asc_desc.size() <= static_cast<size_t>(left_table.num_columns()), "Too many values in asc_desc");
+    CUDF_EXPECTS(key_cols.size() == asc_desc.size(), "Mismatched size between key_cols and asc_desc");
+
+
+    // TODO: replace / drop;
+    //not clear yet...
+    //{
+    auto gdf_col_deleter = [](gdf_column *col) {
+        gdf_column_free(col);
+        delete col;
+    };
+    using gdf_col_ptr = typename std::unique_ptr<gdf_column, decltype(gdf_col_deleter)>;
+    std::vector<gdf_col_ptr> temp_columns_to_free;
+    std::vector<gdf_column*> left_cols_sync(const_cast<gdf_column**>(left_table.begin()), const_cast<gdf_column**>(left_table.end()));
+    std::vector<gdf_column*> right_cols_sync(const_cast<gdf_column**>(right_table.begin()), const_cast<gdf_column**>(right_table.end()));
+    //}
+    
+    for (cudf::size_type i = 0; i < left_table.num_columns(); i++) {
+        //old:
+        //
+        //gdf_column * left_col = const_cast<gdf_column*>(left_table.get_column(i));
+        //gdf_column * right_col = const_cast<gdf_column*>(right_table.get_column(i));
+        //
+        //new:
+        //
+        const auto& left_col = left_table.column(i);
+        const auto& right_col= right_table.column(i);
+
+        //old:
+        //
+        // if (left_col->dtype != GDF_STRING_CATEGORY){
+        //     continue;
+        // }
+        //
+        //new:
+        //
+        if( left_col.type().id() != CATEGORY )
```

Comment:
I think STRING is what you are looking for here...

Reply:
OK. Just for the record, the comment for `gdf_dtype` in `types.h` for `GDF_STRING_CATEGORY`, says: "///< Stores indices of an NVCategory in data and in extra col info a reference to the nv_category".

While, for `STRING`, in `types.hpp:type_id` is "///< String elements". On, the other hand, `CATEGORY` in `type_id` specifies: "///< Categorial/Dictionary type", which seems closer to `GDF_STRING_CATEGORY`s comment. BTW, is there a document with the mapping between the 2 enums?

In any case, I assumed this PR is about API signatures. The implementation is far from ready, is not even compileable, yet.

Reply:
This PR is to define and implement the APIs, as your title says.  We had a couple of meetings discussing strings support. There's a section in the transition guide. https://github.com/rapidsai/cudf/blob/branch-0.11/cpp/docs/TRANSITIONGUIDE.md#strings-support